controller/api/public/grpc_server_test.go

```diff
@@ -186,6 +189,128 @@ spec:
 			}
 		}
 	})
+
+	t.Run("Fails for both resource and namespace set in request", func(t *testing.T) {
```

Comment:
Hooray for tests! Since the existing test for the `ListPods` endpoint is already setup to use a [table driven test](https://github.com/golang/go/wiki/TableDrivenTests), I'd prefer to add new test cases to the table, instead of new tests entirely. Unfortunately, however, the table driven test setup is a bit lacking for the tests that you're adding. But you should feel free to tweak it to satisfy your use case. For instance, to convert this specific test to a table driven test, I'd do:

```diff
diff --git a/controller/api/public/grpc_server_test.go b/controller/api/public/grpc_server_test.go
index 969e5f23..0708e93b 100644
--- a/controller/api/public/grpc_server_test.go
+++ b/controller/api/public/grpc_server_test.go
@@ -3,6 +3,7 @@ package public
 import (
 	"context"
 	"fmt"
+	"reflect"
 	"sort"
 	"strings"
 	"testing"
@@ -19,7 +20,8 @@ type listPodsExpected struct {
 	err     error
 	k8sRes  []string
 	promRes model.Value
-	res     pb.ListPodsResponse
+	req     *pb.ListPodsRequest
+	res     *pb.ListPodsResponse
 }
 
 type listServicesExpected struct {
@@ -42,7 +44,11 @@ func (bs ByService) Len() int           { return len(bs) }
 func (bs ByService) Swap(i, j int)      { bs[i], bs[j] = bs[j], bs[i] }
 func (bs ByService) Less(i, j int) bool { return bs[i].Name <= bs[j].Name }
 
-func listPodResponsesEqual(a pb.ListPodsResponse, b pb.ListPodsResponse) bool {
+func listPodResponsesEqual(a *pb.ListPodsResponse, b *pb.ListPodsResponse) bool {
+	if a == nil || b == nil {
+		return a == b
+	}
+
 	if len(a.Pods) != len(b.Pods) {
 		return false
 	}
@@ -72,7 +78,7 @@ func listPodResponsesEqual(a pb.ListPodsResponse, b pb.ListPodsResponse) bool {
 }
 
 func TestListPods(t *testing.T) {
-	t.Run("Successfully performs a query based on resource type", func(t *testing.T) {
+	t.Run("Queries to the ListPods endpoint", func(t *testing.T) {
 		expectations := []listPodsExpected{
 			listPodsExpected{
 				err: nil,
@@ -142,7 +148,8 @@ spec:
       pod-template-hash: hash-not-meshed
 `,
 				},
-				res: pb.ListPodsResponse{
+				req: &pb.ListPodsRequest{},
+				res: &pb.ListPodsResponse{
 					Pods: []*pb.Pod{
 						&pb.Pod{
 							Name:            "emojivoto/emojivoto-meshed",
@@ -161,6 +168,25 @@ spec:
 					},
 				},
 			},
+			listPodsExpected{
+				err: fmt.Errorf("cannot set both namespace and resource in the request. These are mutually exclusive"),
+				promRes: model.Vector{
+					&model.Sample{
+						Metric:    model.Metric{"pod": "emojivoto-meshed"},
+						Timestamp: 456,
+					},
+				},
+				k8sRes: []string{},
+				req: &pb.ListPodsRequest{
+					Namespace: "test",
+					Selector: &pb.ResourceSelection{
+						Resource: &pb.Resource{
+							Type: pkgK8s.Pod,
+						},
+					},
+				},
+				res: nil,
+			},
 		}
 
 		for _, exp := range expectations {
@@ -179,52 +205,17 @@ spec:
 
 			k8sAPI.Sync()
 
-			rsp, err := fakeGrpcServer.ListPods(context.TODO(), &pb.ListPodsRequest{})
-			if err != exp.err {
+			rsp, err := fakeGrpcServer.ListPods(context.TODO(), exp.req)
+			if !reflect.DeepEqual(err, exp.err) {
 				t.Fatalf("Expected error: %s, Got: %s", exp.err, err)
 			}
 
-			if !listPodResponsesEqual(exp.res, *rsp) {
-				t.Fatalf("Expected: %+v, Got: %+v", &exp.res, rsp)
+			if !listPodResponsesEqual(exp.res, rsp) {
+				t.Fatalf("Expected: %+v, Got: %+v", exp.res, rsp)
 			}
 		}
 	})
 
-	t.Run("Fails for both resource and namespace set in request", func(t *testing.T) {
-		k8sAPI, err := k8s.NewFakeAPI("", make([]string, 0)...)
-		if err != nil {
-			t.Fatalf("NewFakeAPI returned an error: %s", err)
-		}
-
-		fakeGrpcServer := newGrpcServer(
-			&mockProm{Res: model.Vector{
-				&model.Sample{
-					Metric:    model.Metric{"pod": "emojivoto-meshed"},
-					Timestamp: 456,
-				},
-			}},
-			tap.NewTapClient(nil),
-			k8sAPI,
-			"linkerd",
-			[]string{},
-		)
-
-		k8sAPI.Sync()
-
-		_, err = fakeGrpcServer.ListPods(context.TODO(), &pb.ListPodsRequest{
-			Namespace: "test",
-			Selector: &pb.ResourceSelection{
-				Resource: &pb.Resource{
-					Type: pkgK8s.Pod,
-				},
-			},
-		})
-		expectedError := fmt.Errorf("cannot set both namespace and resource in the request. These are mutually exclusive")
-		if err.Error() != expectedError.Error() {
-			t.Fatalf("Expected error: %s, Got: %s", expectedError, err)
-		}
-	})
-
 	t.Run("Reads namespace from resource namespace field", func(t *testing.T) {
 		k8sAPI, err := k8s.NewFakeAPI("", make([]string, 0)...)
 		if err != nil {
```

Hopefully with a few more modifications to that setup, we could satisfy the two other tests that you're adding as well.